File: go/pkg/memory/conversation.go
package memory

import (
	"context"
	"errors"
	"strconv"

	"github.com/haivivi/giztoy/go/pkg/kv"
	"github.com/haivivi/giztoy/go/pkg/recall"
	"github.com/vmihailenco/msgpack/v5"
)

// Conversation is an active dialogue tied to a device or session.
// It stores messages in KV (short-term memory) and provides access to
// recent memory segments via the underlying recall index.
//
// Messages are keyed by nanosecond timestamp for chronological ordering.
// Revert removes the most recent model response and the user message that
// triggered it, enabling a "regenerate" flow.
//
// If a [CompressPolicy] is configured on the parent [Memory], appending
// messages automatically triggers compression when the accumulated message
// count or character count reaches the configured thresholds.
type Conversation struct {
	mem    *Memory // parent memory (for store, index, compressor, policy)
	convID string
	labels []string

	// Auto-compression state: tracks uncompressed messages.
	pendingChars    int
	pendingMsgs     int
	lastCompressErr error // last auto-compression error, if any
}

func newConversation(mem *Memory, convID string, labels []string) *Conversation {
	return &Conversation{
		mem:    mem,
		convID: convID,
		labels: labels,
	}
}

// ID returns the conversation identifier (typically a device or session ID).
func (c *Conversation) ID() string { return c.convID }

// Labels returns the entity labels associated with this conversation
// (e.g., ["person:Alice"]).
func (c *Conversation) Labels() []string { return c.labels }

// LastCompressErr returns the error from the most recent auto-compression
// attempt, or nil if the last attempt succeeded (or no attempt was made).
// This is informational — auto-compression failures do not block [Append].
func (c *Conversation) LastCompressErr() error { return c.lastCompressErr }

// Append stores a message in the conversation.
// If msg.Timestamp is zero, it is set to the current time.
//
// For user messages, a revert point is saved so that [Revert] can undo
// back to this point.
//
// If auto-compression is enabled (a [Compressor] and [CompressPolicy] are
// configured), Append checks whether the accumulated message count or
// character count has reached the policy thresholds. When triggered,
// compression runs synchronously before Append returns.
func (c *Conversation) Append(ctx context.Context, msg Message) error {
	if msg.Timestamp == 0 {
		msg.Timestamp = nowNano()
	}

	data, err := msgpack.Marshal(msg)
	if err != nil {
		return err
	}

	store := c.mem.store
	mid := c.mem.id

	key := convMsgKey(mid, c.convID, msg.Timestamp)
	if err := store.Set(ctx, key, data); err != nil {
		return err
	}

	// Save revert point on user messages.
	if msg.Role == RoleUser {
		rk := convRevertKey(mid, c.convID)
		ts := strconv.FormatInt(msg.Timestamp, 10)
		if err := store.Set(ctx, rk, []byte(ts)); err != nil {
			return err
		}
	}

	// Track pending content for auto-compression.
	c.pendingChars += len(msg.Content)
	c.pendingMsgs++

	// Auto-compress if policy thresholds are reached.
	// On failure, log and continue — messages are safely stored in KV and
	// will be included in the next compression attempt.
	if c.mem.compressor != nil && c.mem.policy.shouldCompress(c.pendingChars, c.pendingMsgs) {
		if err := c.mem.Compress(ctx, c, nil); err != nil {
			c.lastCompressErr = err
		} else {
			c.pendingChars = 0
			c.pendingMsgs = 0
			c.lastCompressErr = nil
		}
	}

	return nil
}

// Recent returns the n most recent messages in chronological order
// (oldest first). If fewer than n messages exist, all are returned.
func (c *Conversation) Recent(ctx context.Context, n int) ([]Message, error) {
	if n <= 0 {
		return nil, nil
	}

	prefix := convMsgPrefix(c.mem.id, c.convID)
	var all []Message

	for entry, err := range c.mem.store.List(ctx, prefix) {
		if err != nil {
			return nil, err
		}
		var msg Message
		if err := msgpack.Unmarshal(entry.Value, &msg); err != nil {
			continue
		}
		all = append(all, msg)
	}

	// KV list is ascending by key (chronological). Take the last n.
	if len(all) > n {
		all = all[len(all)-n:]
	}
	return all, nil
}

// Count returns the total number of messages in the conversation.
func (c *Conversation) Count(ctx context.Context) (int, error) {
	prefix := convMsgPrefix(c.mem.id, c.convID)
	count := 0
	for _, err := range c.mem.store.List(ctx, prefix) {
		if err != nil {
			return 0, err
		}
		count++
	}
	return count, nil
}

// Revert removes the most recent model response(s) and the user message
// that triggered them. This enables a "regenerate last response" flow.
//
// The revert point is the timestamp of the last user message. All messages
// at or after this timestamp are deleted. Returns nil if no revert point
// exists (no user messages have been sent).
func (c *Conversation) Revert(ctx context.Context) error {
	store := c.mem.store
	mid := c.mem.id

	rk := convRevertKey(mid, c.convID)
	data, err := store.Get(ctx, rk)
	if err != nil {
		if errors.Is(err, kv.ErrNotFound) {
			return nil // nothing to revert
		}
		return err
	}

	revertTS, err := strconv.ParseInt(string(data), 10, 64)
	if err != nil {
		return err
	}

	// Collect keys to delete: all messages with timestamp >= revertTS.
	prefix := convMsgPrefix(mid, c.convID)
	var toDelete []kv.Key

	for entry, err := range store.List(ctx, prefix) {
		if err != nil {
			return err
		}
		// Extract timestamp from the last key segment.
		tsStr := entry.Key[len(entry.Key)-1]
		ts, err := strconv.ParseInt(tsStr, 10, 64)
		if err != nil {
			continue
		}
		if ts >= revertTS {
			toDelete = append(toDelete, entry.Key)
		}
	}

	if len(toDelete) == 0 {
		return nil
	}

	// Delete the messages.
	if err := store.BatchDelete(ctx, toDelete); err != nil {
		return err
	}

	// Update the revert point to the previous user message.
	// Scan remaining messages to find the latest user message.
	var latestUserTS int64
	for entry, err := range store.List(ctx, prefix) {
		if err != nil {
			return err
		}
		var msg Message
		if err := msgpack.Unmarshal(entry.Value, &msg); err != nil {
			continue
		}
		if msg.Role == RoleUser && msg.Timestamp > latestUserTS {
			latestUserTS = msg.Timestamp
		}
	}

	if latestUserTS > 0 {
		ts := strconv.FormatInt(latestUserTS, 10)
		return store.Set(ctx, rk, []byte(ts))
	}
	// No user messages remain; delete the revert key.
	return store.Delete(ctx, rk)
}

// RecentSegments returns the n most recent memory segments from the
// underlying recall index. These complement [Recent] messages to give
// the LLM additional historical context beyond the conversation window.
func (c *Conversation) RecentSegments(ctx context.Context, n int) ([]recall.Segment, error) {
	return c.mem.index.RecentSegments(ctx, n)
}

// All returns all messages in chronological order. This is used by the
// compression pipeline to read all messages before compressing them
// into segments.
func (c *Conversation) All(ctx context.Context) ([]Message, error) {
	prefix := convMsgPrefix(c.mem.id, c.convID)
	var msgs []Message

	for entry, err := range c.mem.store.List(ctx, prefix) {
		if err != nil {
			return nil, err
		}
		var msg Message
		if err := msgpack.Unmarshal(entry.Value, &msg); err != nil {
			continue
		}
		msgs = append(msgs, msg)
	}
	return msgs, nil
}

// Clear removes all messages and the revert point for this conversation.
// It also resets the auto-compression counters.
func (c *Conversation) Clear(ctx context.Context) error {
	mid := c.mem.id
	prefix := convMsgPrefix(mid, c.convID)
	var keys []kv.Key

	for entry, err := range c.mem.store.List(ctx, prefix) {
		if err != nil {
			return err
		}
		keys = append(keys, entry.Key)
	}

	// Also delete the revert key.
	keys = append(keys, convRevertKey(mid, c.convID))

	if len(keys) > 0 {
		if err := c.mem.store.BatchDelete(ctx, keys); err != nil {
			return err
		}
	}

	c.pendingChars = 0
	c.pendingMsgs = 0
	return nil
}

File: rust/memory/src/conversation.rs
use giztoy_recall::Segment;

use crate::error::MemoryError;
use crate::keys::{conv_msg_key, conv_msg_prefix, conv_revert_key};
use crate::memory::Memory;
use crate::types::{Message, Role, now_nano};

/// Active dialogue session tied to a device or session ID.
/// Messages are stored in KV keyed by nanosecond timestamp for ordering.
pub struct Conversation<'m> {
    mem: &'m Memory,
    conv_id: String,
    labels: Vec<String>,

    pending_chars: usize,
    pending_msgs: usize,
    last_compress_err: Option<MemoryError>,
}

impl<'m> Conversation<'m> {
    pub(crate) fn new(mem: &'m Memory, conv_id: String, labels: Vec<String>) -> Self {
        Self {
            mem,
            conv_id,
            labels,
            pending_chars: 0,
            pending_msgs: 0,
            last_compress_err: None,
        }
    }

    pub fn id(&self) -> &str { &self.conv_id }

    pub fn labels(&self) -> &[String] { &self.labels }

    pub fn last_compress_err(&self) -> Option<&MemoryError> {
        self.last_compress_err.as_ref()
    }

    /// Store a message. Auto-fills timestamp if zero.
    /// Saves revert point on user messages.
    /// Triggers auto-compression when policy thresholds are reached.
    pub async fn append(&mut self, mut msg: Message) -> Result<(), MemoryError> {
        if msg.timestamp == 0 {
            msg.timestamp = now_nano();
        }

        let data = rmp_serde::to_vec_named(&msg)
            .map_err(|e| MemoryError::Serialization(e.to_string()))?;

        let store = self.mem.kv_store();
        let mid = self.mem.id();

        let key = conv_msg_key(mid, &self.conv_id, msg.timestamp);
        store.set(&key, &data)?;

        if msg.role == Role::User {
            let rk = conv_revert_key(mid, &self.conv_id);
            let ts_str = msg.timestamp.to_string();
            store.set(&rk, ts_str.as_bytes())?;
        }

        self.pending_chars += msg.content.len();
        self.pending_msgs += 1;

        if self.mem.has_compressor()
            && self.mem.policy().should_compress(self.pending_chars, self.pending_msgs)
        {
            match self.mem.compress(self, None).await {
                Ok(()) => {
                    self.pending_chars = 0;
                    self.pending_msgs = 0;
                    self.last_compress_err = None;

                    if let Err(e) = self.mem.compact().await {
                        self.last_compress_err = Some(e);
                    }
                }
                Err(e) => {
                    self.last_compress_err = Some(e);
                }
            }
        }

        Ok(())
    }

    /// Return the n most recent messages in chronological order (oldest first).
    pub fn recent(&self, n: usize) -> Result<Vec<Message>, MemoryError> {
        if n == 0 {
            return Ok(vec![]);
        }

        let all = self.scan_messages()?;
        let start = if all.len() > n { all.len() - n } else { 0 };
        Ok(all[start..].to_vec())
    }

    /// Return total message count.
    pub fn count(&self) -> Result<usize, MemoryError> {
        let prefix = conv_msg_prefix(self.mem.id(), &self.conv_id);
        let entries = self.mem.kv_store().scan(&prefix)?;
        Ok(entries.len())
    }

    /// Revert: delete the last user message and all subsequent model replies.
    pub fn revert(&self) -> Result<(), MemoryError> {
        let store = self.mem.kv_store();
        let mid = self.mem.id();

        let rk = conv_revert_key(mid, &self.conv_id);
        let data = match store.get(&rk)? {
            Some(d) => d,
            None => return Ok(()),
        };

        let revert_ts: i64 = String::from_utf8(data)
            .map_err(|e| MemoryError::Serialization(e.to_string()))?
            .parse()
            .map_err(|e: std::num::ParseIntError| MemoryError::Serialization(e.to_string()))?;

        let prefix = conv_msg_prefix(mid, &self.conv_id);
        let entries = store.scan(&prefix)?;

        let mut to_delete: Vec<String> = Vec::new();
        for (key, _) in &entries {
            let ts_str = match key.rsplit(':').next() {
                Some(s) => s,
                None => continue,
            };
            let ts: i64 = match ts_str.parse() {
                Ok(t) => t,
                Err(_) => continue,
            };
            if ts >= revert_ts {
                to_delete.push(key.clone());
            }
        }

        if to_delete.is_empty() {
            return Ok(());
        }

        let key_refs: Vec<&str> = to_delete.iter().map(|s| s.as_str()).collect();
        store.batch_delete(&key_refs)?;

        // Find the new latest user message for the updated revert point.
        let remaining = store.scan(&prefix)?;
        let mut latest_user_ts: i64 = 0;
        for (_, value) in &remaining {
            if let Ok(msg) = rmp_serde::from_slice::<Message>(value)
                && msg.role == Role::User
                && msg.timestamp > latest_user_ts
            {
                latest_user_ts = msg.timestamp;
            }
        }

        if latest_user_ts > 0 {
            let ts_str = latest_user_ts.to_string();
            store.set(&rk, ts_str.as_bytes())?;
        } else {
            store.delete(&rk)?;
        }

        Ok(())
    }

    /// Return the n most recent memory segments from the recall index.
    pub fn recent_segments(&self, n: usize) -> Result<Vec<Segment>, MemoryError> {
        Ok(self.mem.index().recent_segments(n)?)
    }

    /// Return all messages in chronological order.
    pub fn all(&self) -> Result<Vec<Message>, MemoryError> {
        self.scan_messages()
    }

    /// Remove all messages and the revert point. Resets pending compression
    /// counters and last compression error in this conversation handle.
    pub fn clear(&mut self) -> Result<(), MemoryError> {
        let store = self.mem.kv_store();
        let mid = self.mem.id();
        let prefix = conv_msg_prefix(mid, &self.conv_id);

        let entries = store.scan(&prefix)?;
        let mut keys: Vec<String> = entries.into_iter().map(|(k, _)| k).collect();
        keys.push(conv_revert_key(mid, &self.conv_id));

        if !keys.is_empty() {
            let key_refs: Vec<&str> = keys.iter().map(|s| s.as_str()).collect();
            store.batch_delete(&key_refs)?;
        }

        self.pending_chars = 0;
        self.pending_msgs = 0;
        self.last_compress_err = None;

        Ok(())
    }

    fn scan_messages(&self) -> Result<Vec<Message>, MemoryError> {
        let prefix = conv_msg_prefix(self.mem.id(), &self.conv_id);
        let entries = self.mem.kv_store().scan(&prefix)?;

        let mut msgs = Vec::new();
        for (_, value) in entries {
            match rmp_serde::from_slice::<Message>(&value) {
                Ok(msg) => msgs.push(msg),
                Err(_) => continue,
            }
        }
        Ok(msgs)
    }
}
